### PR TITLE
FIX: Extend ref pin search outside component with create port on component

### PR DESCRIPTION
**Pull request by svandenb-dev**

### CODE REVIEW

**Review by ring630 — APPROVED** [review]
LGTM

File: src/pyedb/dotnet/edb_core/components.py [hunk 1]
```diff
@@ -766,20 +766,44 @@ def create_port_on_pins(self, refdes, pins, reference_pins, impedance=50.0, port
...
         if isinstance(reference_pins, str):
             reference_pins = [reference_pins]
-        if isinstance(refdes, str) or isinstance(refdes, EDBComponent):
+        if isinstance(reference_pins, list):
+            _temp = []
+            for ref_pin in reference_pins:
+                if isinstance(ref_pin, int):
+                    if ref_pin in self._padstack.instances:
+                        _temp.append(self._padstack.instances[ref_pin])
+                if isinstance(ref_pin, str):
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:781]:
```suggestion
                elif isinstance(ref_pin, str):
```
Same comment as before

File: src/pyedb/dotnet/edb_core/components.py [hunk 2]
```diff
@@ -766,20 +766,44 @@ def create_port_on_pins(self, refdes, pins, reference_pins, impedance=50.0, port
...
+                    if ref_pin in self.instances[refdes].pins:
+                        _temp.append(self.instances[refdes].pins[ref_pin])
+                    else:
+                        p = [pp for pp in list(self._padstack.instances.values()) if pp.name == ref_pin]
+                        if p:
+                            _temp.append(p)
+                if isinstance(ref_pin, EDBPadstackInstance):
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:788]:
```suggestion
                elif isinstance(ref_pin, EDBPadstackInstance):
```
Same comment as before

File: src/pyedb/dotnet/edb_core/components.py [hunk 3]
```diff
@@ -766,20 +766,44 @@ def create_port_on_pins(self, refdes, pins, reference_pins, impedance=50.0, port
...
+                    _temp.append(ref_pin.name)
+            reference_pins = _temp
+        if isinstance(reference_pins, int):
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:791]:
```suggestion
        elif isinstance(reference_pins, int):
```
Same comment as before

File: src/pyedb/dotnet/edb_core/components.py [hunk 4]
```diff
@@ -766,20 +766,44 @@ def create_port_on_pins(self, refdes, pins, reference_pins, impedance=50.0, port
...
+            if reference_pins in self._padstack.instances:
+                reference_pins = self._padstack.instances[reference_pins]
+        if isinstance(refdes, str):
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:794]:
```suggestion
        elif isinstance(refdes, str):
```
Same comment as before

File: src/pyedb/dotnet/edb_core/components.py [hunk 5]
```diff
@@ -766,20 +766,44 @@ def create_port_on_pins(self, refdes, pins, reference_pins, impedance=50.0, port
...
             refdes = self.instances[refdes]
+        if isinstance(refdes, self._pedb._edb.Cell.Hierarchy.Component):
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:796]:
```suggestion
        elif isinstance(refdes, self._pedb._edb.Cell.Hierarchy.Component):
```
Same comment as before

File: src/pyedb/dotnet/edb_core/components.py [hunk 6]
```diff
@@ -793,15 +817,14 @@ def create_port_on_pins(self, refdes, pins, reference_pins, impedance=50.0, port
...
-                    ref_cmp_pins.append(cmp_ref_pin[0])
+                if ref_pin_name in refdes_pins:
+                    ref_cmp_pins.append(refdes_pins[ref_pin_name])
+                elif "-" in ref_pin_name and ref_pin_name.split("-")[1] in refdes_pins:
+                    ref_cmp_pins.append(refdes_pins[ref_pin_name.split("-")[1]])
             if not ref_cmp_pins:
                 return
             reference_pins = ref_cmp_pins
-        if not len([pin for pin in reference_pins if isinstance(pin, EDBPadstackInstance)]) == len(reference_pins):
+        if not reference_pins:
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:827]: You might considering logging a warning/error here

File: src/pyedb/dotnet/edb_core/components.py [hunk 7]
```diff
@@ -981,16 +1016,23 @@ def create_port_on_component(
             for p in ref_pins:
                 if not p.IsLayoutPin():
                     p.SetIsLayoutPin(True)
-            if len(ref_pins) == 0:
-                self._logger.info("No reference pin found on component {}.".format(component.GetName()))
+            if not ref_pins:
+                self._logger.warning(
+                    "No reference pins found on component, extending pin search at the component "
+                    "vicinity Pingroups are not supported, the closest pin will be selected."
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:1022]:
```suggestion
                    "No reference pins found on component, the closest pin will be selected."
```
The message is not clear to me, what about this one ? Less information but at least I'm not lost when reading it.

File: src/pyedb/dotnet/edb_core/components.py [hunk 8]
```diff
@@ -981,16 +1016,23 @@ def create_port_on_component(
...
                 else:
+                    for pin in ref_pins:
+                        pin.is_pin = True
                     ref_pin_group = self.create_pingroup_from_pins(ref_pins)
                     if not ref_pin_group:
                         return False
-                    ref_pin_group_term = self._create_pin_group_terminal(ref_pin_group, isref=True)
+                    ref_pin_group_term = self._create_pin_group_terminal(ref_pin_group, isref=False)
                     if not ref_pin_group_term:
                         return False
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:1037]: What about logging when we return `False` ?

**Review by SMoraisAnsys — CHANGES_REQUESTED** [review]
I left a couple of comments.

File: src/pyedb/dotnet/edb_core/components.py (hunk 8 above)
Comment (svandenb-dev) [reply src/pyedb/dotnet/edb_core/components.py:1036]: Yes correct

File: src/pyedb/dotnet/edb_core/components.py (hunk 6 above)
Comment (svandenb-dev) [reply src/pyedb/dotnet/edb_core/components.py:827]: Done

File: src/pyedb/dotnet/edb_core/components.py [hunk 9]
```diff
@@ -934,6 +958,18 @@ def create_port_on_component(
...
         pin_layers = cmp_pins[0].GetPadstackDef().GetData().GetLayerNames()
         if port_type == SourceType.CoaxPort:
+            ref_pins = [
+                p
+                for p in list(component.LayoutObjs)
+                if int(p.GetObjType()) == 1 and p.GetNet().GetName() in reference_net
+            ]
+            if not ref_pins:
+                self._logger.error(
+                    "The no reference pins found on component. You might consider"
```
Comment (SMoraisAnsys) [inline src/pyedb/dotnet/edb_core/components.py:968]:
```suggestion
                    "No reference pins found on component. You might consider"
```

**Review by SMoraisAnsys — DISMISSED** [review]
LGTM, I found a last typo though

**Review by SMoraisAnsys — APPROVED** [review]

**Pull request merged**